File: internal/domain/reconcile.go
package domain

import "sort"

// FiltrarExcluidos remove do alvo os commits ja marcados como excluidos no
// lock (§3) - sem isso, todo verificar reportaria o mesmo falso-positivo pra
// sempre.
func FiltrarExcluidos(alvo TargetSet, excluidos []Exclusion) TargetSet {
	excluido := make(map[string]bool, len(excluidos))
	for _, e := range excluidos {
		excluido[e.Commit] = true
	}
	filtrado := make(TargetSet, len(alvo))
	for chamado, tt := range alvo {
		var commits []CommitRef
		for _, c := range tt.Commits {
			if !excluido[c.HashOrigem] {
				commits = append(commits, c)
			}
		}
		filtrado[chamado] = TaskTarget{Chamado: tt.Chamado, Task: tt.Task, Titulo: tt.Titulo, Commits: commits}
	}
	return filtrado
}

// DiffTasks calcula a diferenca simetrica entre alvo e lock (§5, §9).
func DiffTasks(alvo, lockTasks TargetSet) (novas, removidas []string) {
	for chamado := range alvo {
		if _, ok := lockTasks[chamado]; !ok {
			novas = append(novas, chamado)
		}
	}
	for chamado := range lockTasks {
		if _, ok := alvo[chamado]; !ok {
			removidas = append(removidas, chamado)
		}
	}
	sort.Strings(novas)
	sort.Strings(removidas)
	return novas, removidas
}

// Reconciliar cruza as 3 fontes (§2, §9) e produz o VersionStatus. `presentes`
// e `conflitantes` sao pre-computados pelo chamador (services.PresenceOracle e
// GitRepo.PredictMerge) - esta funcao fica pura.
func Reconciliar(alvo TargetSet, lock Lock, presentes map[string]bool, conflitantes []CommitRef) VersionStatus {
	novas, removidas := DiffTasks(alvo, lock.Tasks)

	var faltantes []CommitRef
	for _, tt := range alvo {
		for _, c := range tt.Commits {
			if !presentes[c.HashOrigem] {
				faltantes = append(faltantes, c)
			}
		}
	}

	lockIntegro := true
	var sumidos []string
	for _, tt := range lock.Tasks {
		for _, c := range tt.Commits {
			if !presentes[c.HashOrigem] {
				lockIntegro = false
				sumidos = append(sumidos, c.HashOrigem)
			}
		}
	}
	sort.Strings(sumidos)

	verde := len(novas) == 0 && len(removidas) == 0 && lockIntegro && len(faltantes) == 0

	return VersionStatus{
		Verde:          verde,
		TasksNovas:     novas,
		TasksRemovidas: removidas,
		LockIntegro:    lockIntegro,
		CommitsSumidos: sumidos,
		Faltantes:      faltantes,
		Conflitantes:   conflitantes,
	}
}

File: motor/domain/reconcile.py
"""Cruzamento das tres fontes: Tickio (alvo) x estado x git."""

from __future__ import annotations

from dataclasses import replace

from motor.domain.types import (
    Alvo,
    Atribuicao,
    CommitRef,
    Exclusion,
    Presence,
    TargetSet,
    VersionStatus,
)


def filtrar_excluidos(
    alvo: TargetSet, excluidos: list[Exclusion], versao: str
) -> TargetSet:
    """Tira do alvo os commits marcados como excluidos.

    `versao_numero` None vale para todo o repo (ex.: commit revertido); com
    valor, so para aquela versao. Sem isso, todo verificar reportaria o mesmo
    falso-positivo para sempre.
    """
    fora = {
        e.hash_origem
        for e in excluidos
        if e.versao_numero is None or e.versao_numero == versao
    }
    return {
        chamado: replace(
            tt, commits=[c for c in tt.commits if c.hash_origem not in fora]
        )
        for chamado, tt in alvo.items()
    }


def diff_tasks(
    alvo: TargetSet, anteriores: list[Atribuicao]
) -> tuple[list[str], list[str]]:
    """Diferenca simetrica entre o alvo de agora e o estado gravado no run
    anterior. E o que detecta tarefa desmarcada no Tickio — comparar depois de
    sobrescrever apagaria a propria evidencia.
    """
    antes = {a.chamado for a in anteriores}
    novas = [ch for ch, tt in alvo.items() if ch not in antes and tt.commits]
    removidas = [ch for ch in antes if ch not in alvo]
    return sorted(novas), sorted(removidas)


def atribuicoes_de(
    alvo: TargetSet, presentes: dict[str, Presence]
) -> list[Atribuicao]:
    """Projeta o alvo resolvido no formato que vai para o estado."""
    return [
        Atribuicao(
            chamado=tt.chamado,
            marcada=tt.marcada,
            estado=(
                "aplicado"
                if tt.commits
                and all(
                    presentes.get(c.hash_origem, Presence.AUSENTE) != Presence.AUSENTE
                    for c in tt.commits
                )
                else "pendente"
            ),
            commits=[c.hash_origem for c in tt.commits],
        )
        for tt in alvo.values()
    ]


def reconciliar(
    alvo: Alvo,
    anteriores: list[Atribuicao],
    sem_entrega: dict[str, str],
    presentes: dict[str, Presence],
    conflitantes: list[CommitRef],
    suspeitos_conteudo: list[CommitRef] = (),
) -> VersionStatus:
    """Produz o VersionStatus. Funcao pura: `presentes`, `conflitantes` e
    `suspeitos_conteudo` chegam pre-computados pelo chamador.
    """
    novas, removidas = diff_tasks(alvo.tasks, anteriores)

    faltantes: list[CommitRef] = []
    ancestrais: list[CommitRef] = []
    for tt in alvo.tasks.values():
        for c in tt.commits:
            if presentes.get(c.hash_origem, Presence.AUSENTE) == Presence.AUSENTE:
                faltantes.append(c)
            else:
                ancestrais.append(c)

    # So atribuicao aplicada pode ter commit "sumido": pendente e commit que
    # ainda nao foi cherry-pickado, entao estar ausente do alvo e o normal —
    # contar isso diria "estado divergente do git" (estado corrompido) numa
    # versao recem-criada, mandando o operador investigar o que nao existe.
    sumidos = sorted(
        {
            h
            for a in anteriores
            if a.estado == "aplicado"
            for h in a.commits
            if presentes.get(h, Presence.AUSENTE) == Presence.AUSENTE
        }
    )
    estado_integro = not sumidos

    # Tarefa marcada sem nenhum commit achado nao pode passar despercebida
    # (falso-verde). So sai da lista se reconhecida em sem_entrega.
    sem_commits = sorted(
        ch for ch, tt in alvo.tasks.items() if not tt.commits and ch not in sem_entrega
    )

    verde = (
        not novas
        and not removidas
        and estado_integro
        and not faltantes
        and not sem_commits
        and not alvo.ambiguas
    )

    return VersionStatus(
        verde=verde,
        tasks_novas=novas,
        tasks_removidas=removidas,
        tasks_ambiguas=list(alvo.ambiguas),
        estado_integro=estado_integro,
        commits_sumidos=sumidos,
        faltantes=faltantes,
        ancestrais=ancestrais,
        conflitantes=conflitantes,
        suspeitos_conteudo=list(suspeitos_conteudo),
        tasks_sem_commits=sem_commits,
    )
